File: LeetCode/201-300/211. Design Add and Search Words Data Structure/src/Solution.java
class WordDictionary {

    private TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }

    public void addWord(String word) {
        root.add(word, 0);
    }

    public boolean search(String word) {
        return root.search(word, 0);
    }

    private static class TrieNode {
        private TrieNode[] children;
        private boolean endFlag;

        private TrieNode() {
            children = new TrieNode[26];
            endFlag = false;
        }

        private void add(String word, int index) {
            if (index >= word.length()) {
                this.endFlag = true;
            } else {
                int currInd = word.charAt(index) - 'a';
                if (children[currInd] == null) {
                    children[currInd] = new TrieNode();
                }
                children[currInd].add(word, index + 1);
            }
        }

        private boolean search(String word, int index) {
            if (index == word.length()) {
                return endFlag;
            }
            if (word.charAt(index) == '.') {
                for (int i = 0; i < 26; i++) {
                    if (children[i] != null && children[i].search(word, index + 1)) {
                        return true;
                    }
                }
                return false;
            } else {
                int currInd = word.charAt(index) - 'a';
                return children[currInd] != null && children[currInd].search(word, index + 1);
            }
        }
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: LeetCode/201-300/211. Design Add and Search Words Data Structure/Solution.cpp
class TrieNode {
    vector<TrieNode*> children;
    bool endWord;

public:
    TrieNode() {
        children = vector<TrieNode*>(26, nullptr);
        endWord = false;
    }

    void addWord(string word, int index) {
        if (index == word.size()) {
            endWord = true;
            return;
        }
        int childInd = word[index] - 'a';
        if (children[childInd] == nullptr) {
            children[childInd] = new TrieNode();
        }
        children[childInd]->addWord(word, index + 1);
    }

    bool search(string word, int index) {
        if (index == word.size()) {
            return endWord;
        }
        if (word[index] == '.') {
            for (int i = 0; i < 26; ++i) {
                if (children[i] != nullptr && children[i]->search(word, index + 1)) {
                    return true;
                }
            }
            return false;
        } else {
            int childInd = word[index] - 'a';
            return children[childInd] != nullptr && children[childInd]->search(word, index + 1);
        }
    }
};

class WordDictionary {
    TrieNode* root;

public:
    WordDictionary() {
        root = new TrieNode();
    }

    void addWord(string word) {
        root->addWord(word, 0);
    }

    bool search(string word) {
        return root->search(word, 0);
    }
};